### PR TITLE
Added PluginAPI hooks for message_move and message_delete

### PR DESCRIPTION
**Pull request by JE4GLE**
This pull request implements the requested PluginAPI Hooks from ticket #9499.

_The wiki needs to be updated, once the change is released._

### message_delete

Triggered when a message is deleted.

_Arguments:_
  * uids: _The messages uids_
  * mailbox: _The mailbox the message is deleted from_

_Return values:_
  * abort: set to true to skip the delete and fail

### message_move

Triggered when a message is moved from one folder to another.
The message is identified by the `uid`. The `mailbox` and `target` value may be equal, if it is moved to the same folder.

_Arguments:_
  * uids: _The messages uids_
  * source_folder: _The mailbox the message is moved from_
  * target_folder: _The mailbox the message is moved to_

_Return values:_
  * abort: set to true to skip the move and fail

### CODE REVIEW

**Comment by alecpl** [comment]
I think better place to put these will be `rcube_imap`. Also, for your use-case this might be enough, but maybe we should take the `abort` flag into consideration.

**Comment by alecpl** [comment]
I guess it depends what you want to do with it. For example `message_saved` is in `rcube_imap`. It's obvious why it makes sense to keep these closer to the storage class. `rcube_imap::move_message()` might be called from various places, e.g. `folder_purge` action uses it too. So, you can handle different move message actions from a single place.

Putting these in actions would make sense if you wanted to interfere in any way with the UI, I suppose.

**Comment by JE4GLE** [comment]
Thank you, I adjusted the code accordingly!

File: program/lib/Roundcube/rcube_imap.php [hunk 1]
```diff
@@ -2589,6 +2589,15 @@ public function move_message($uids, $to_mbox, $from_mbox = '')
             return false;
         }
 
+        $abortedUids = [];
+        foreach ($uids as $uid) {
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_imap.php:2593]: `$uids` is a string here.

File: program/lib/Roundcube/rcube_imap.php [hunk 2]
```diff
@@ -2590,13 +2590,15 @@ public function move_message($uids, $to_mbox, $from_mbox = '')
         }
 
         $abortedUids = [];
-        foreach ($uids as $uid) {
+        $uidsArray = explode(',', $uids);
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_imap.php:2593]: Nope. $uids can be "*" or "1:*" or "1-2". I suggest to skip the loop and pass $uids as is to the hook. No aborted uids. Just return (false or []) on abort.

File: program/lib/Roundcube/rcube_imap.php [hunk 3]
```diff
@@ -2699,6 +2704,11 @@ public function delete_message($uids, $folder = null)
             return false;
         }
 
+        $plugin = $this->plugins->exec_hook('message_delete', ['mailbox' => $folder, 'uids' => $uids]);
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_imap.php:2707]: I know for example `message_read` hook has 'mailbox' argument, but I prefer 'folder' that is used in other hooks.

File: program/lib/Roundcube/rcube_imap.php [hunk 4]
```diff
@@ -2589,6 +2589,11 @@ public function move_message($uids, $to_mbox, $from_mbox = '')
             return false;
         }
 
+        $plugin = $this->plugins->exec_hook('message_move', ['mailbox' => $from_mbox, 'target' => $to_mbox, 'uids' => $uids]);
+        if (isset($plugin['abort']) && $plugin['abort']) {
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_imap.php:2593]: No need to use isset(), 'abort' is always set.

File: program/lib/Roundcube/rcube_imap.php [hunk 5]
```diff
@@ -2589,6 +2589,11 @@ public function move_message($uids, $to_mbox, $from_mbox = '')
             return false;
         }
 
+        $plugin = $this->plugins->exec_hook('message_move', ['mailbox' => $from_mbox, 'target' => $to_mbox, 'uids' => $uids]);
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_imap.php:2592]: How about 'from' and 'to' or 'source_folder' and 'target_folder'?

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/lib/Roundcube/rcube_imap.php [hunk 6]
```diff
@@ -2589,6 +2589,11 @@ public function move_message($uids, $to_mbox, $from_mbox = '')
             return false;
         }
 
+        $plugin = $this->plugins->exec_hook('message_move', ['from_folder' => $from_mbox, 'target_folder' => $to_mbox, 'uids' => $uids]);
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_imap.php:2592]: All right, we're almost there. If we use 'target_folder' then we should use 'source_folder' or just 'folder' and 'target'. Or even just 'from' and 'to' like in the variable names.

**Comment by JE4GLE** [comment]
Thank you for all your helpful comments! I hope the PR can be merged quickly :)

**Pull request merged**